java/arcs/core/data/CapabilityNew.kt

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data
+
+import arcs.core.util.Time
+
+sealed class CapabilityNew {
+    enum class Comparison { LessStrict, Equivalent, Stricter }
+
+    fun isEquivalent(other: CapabilityNew): Boolean {
+        return compare(other) == Comparison.Equivalent
+    }
+    fun contains(other: CapabilityNew) = isEquivalent(other)
+    fun isLessStrict(other: CapabilityNew) = compare(other) == Comparison.LessStrict
+    fun isSameOrLessStrict(other: CapabilityNew) = compare(other) != Comparison.Stricter
+    fun isStricter(other: CapabilityNew) = compare(other) == Comparison.Stricter
+    fun isSameOrStricter(other: CapabilityNew) = compare(other) != Comparison.LessStrict
+
+    abstract fun compare(other: CapabilityNew): Comparison
```

Comment:
You could also define this as follows, which will throw errors when different types are compared:
```suggestion
fun compare(other: CapabilityNew): Comparison {
   when (this) {
       is Persistence -> compare(other as Persistence)
       is TtlNew -> compare(other as TtlNew)
   }
}
```

and in each subclass define an approprate compare method. e.g., 
```
data class Persistence(...) {
   fun compare(other: Persistence): Comparison {...}
}
```

Reply:
i prefer the way it is (unless it's anti-kotlin for some reason), because it is more obvious that every new capability that is supported needs to override `compare`.

Reply:
OK to leave it as is:
> every new capability that is supported needs to override compare
The when clause would force us to add a compare operator.

Also, do you want to prevent comparing two different types altogether?

Reply:
Also, there is a way to enforce compatible type in the compare methods. See https://github.com/PolymerLabs/arcs/blob/d1fa05d6c0c5dd111431e9a6bfb92db9bb95e86a/java/arcs/core/analysis/AbstractValue.kt#L39

https://github.com/PolymerLabs/arcs/blob/d1fa05d6c0c5dd111431e9a6bfb92db9bb95e86a/java/arcs/core/analysis/AbstractValue.kt#L62

and

a corresponding implementation of `join`:

https://github.com/PolymerLabs/arcs/blob/5feb9389c6a4f361a810d0be95baa0335caacb20/java/arcs/core/analysis/AbstractSet.kt#L19

Reply:
thanks! I did it this way, but then I just realized that the original way was better.
the `return when (this) { 
is Persistence->...
...
`
enforces all types are actually listed, because there is no else and `when` becomes non exhaustive if a new capability is added.
(otoh when i did it with a similar to AbstractValue pattern, it will pass compilation, but fail the test with infinite loop).

Reply:
Thanks for humoring me. It would have been OK with your implementation too. :)

Also, I think you will need a `else` clause that fails in `Capability.compare` method. As far as I know, the Kotlin compiler only issues a  warning (and not an error) if the list is non-exhaustive. 

Yeah, for the `AbstractValue` trick, you will have needed two different names: `Capability.compare()` for the non-abstract method and `Capability.compareImpl` for the abstract method.

Reply:
I double checked now - the non exhaustive is definitely a compile time error, so i'll leave it as is for now and we can always revisit and tweak later. Thanks!

Reply:
Thanks for checking. That is good to know.